lib/utils.ts

```diff
@@ -78,7 +90,28 @@ function mergeObject<TObject extends Record<string, unknown>>(target: TObject |
                 // so that we can still use "fastMerge" to merge the source value,
                 // to ensure that nested null values are removed from the merged object.
                 const targetValueWithFallback = (targetValue ?? {}) as TObject;
-                destination[key] = fastMerge(targetValueWithFallback, sourceValue, shouldRemoveNestedNulls);
+
+                // If we are batching merge changes and the previous merge change (targetValue) is null,
+                // it means we want to fully replace this object when merging the batched changes with the Onyx value.
+                // To achieve this, we first mark these nested objects with an internal flag. With the desired objects
+                // marked, when calling this method again with "shouldReplaceMarkedObjects" set to true we can proceed
+                // to effectively replace them in the next condition.
+                if (isBatchingMergeChanges && targetValue === null) {
+                    (targetValueWithFallback as Record<string, unknown>)[ONYX_INTERNALS__REPLACE_OBJECT_MARK] = true;
+                }
+
+                // Then, when merging the batched changes with the Onyx value, if a nested object of the batched changes
+                // has the internal flag set, we replace the entire destination object with the source one and remove
+                // the flag.
+                if (shouldReplaceMarkedObjects && sourceValue[ONYX_INTERNALS__REPLACE_OBJECT_MARK] === true) {
```

Comment:
Could we do something like this?

```suggestion
                if (shouldReplaceMarkedObjects && sourceValue[ONYX_INTERNALS__REPLACE_OBJECT_MARK]) {
```